File: src/leetcode980.java
public class leetcode980 {
    private int[][] grid;
    private int R, C;
    private int visited;
    private int start, end;

    private int[][] dirs = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

    public int uniquePathsIII(int[][] grid) {
        this.grid = grid;
        R = grid.length;
        C = grid[0].length;
        visited = 0;

        int left = R * C;

        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                if (grid[i][j] == 1) {
                    start = i * C + j;
                    grid[i][j] = 0;
                } else if (grid[i][j] == 2) {
                    end = i * C + j;
                    grid[i][j] = 0;
                } else if (grid[i][j] == -1) {
                    left--;
                }

            }
        }
        return dfs(start, left);
    }

    private int dfs(int v, int left) {
        int x = v / C, y = v % C;
        visited += (1 << v);
        left--;
        if (left == 0 && v == end) {
            visited -= (1 << v);
            return 1;
        }

        int res = 0;
        for (int d = 0; d < 4; d++) {
            int nextx = x + dirs[d][0], nexty = y + dirs[d][1];

            int next = nextx * C + nexty;

            if (inArea(nextx, nexty) && grid[nextx][nexty] == 0 && (visited & (1 << next)) == 0)
                res += dfs(nextx * C + nexty, left);
        }
        visited -= (1 << v);
        return res;

    }

    private boolean inArea(int x, int y) {
        return x >= 0 && x < R && y >= 0 && y < C;
    }
}

File: src/Chapter08HamiltonLoop/python/leetcode980.py
class Solution:
    steps = [[0, 1], [1, 0], [0, -1], [-1, 0]]

    def uniquePathsIII(self, grid):
        self.graph = grid
        self.visited = [[False for _ in range(len(grid[i]))] for i in range(len(grid))]
        start_i = -1
        start_j = -1
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == -1:
                    self.visited[i][j] = True
                elif grid[i][j] == 1:
                    start_i = i
                    start_j = j
        if start_i == -1:
            return 0
        return self.dfs(start_i, start_j)

    def dfs(self, i, j):
        self.visited[i][j] = True

        if self.graph[i][j] == 2 and all([all(self.visited[i]) for i in range(len(self.visited))]):
            self.visited[i][j] = False
            return 1
        res = 0
        for step in self.steps:
            next_x = i + step[0]
            next_y = j + step[1]
            if next_x >= 0 and next_x < len(self.graph) and next_y >= 0 and next_y < len(self.graph[next_x]) and not \
            self.visited[next_x][next_y]:
                res += self.dfs(next_x, next_y)
        self.visited[i][j] = False
        return res

